examples/keras_recipes/distillation_recipe.py

```diff
@@ -0,0 +1,518 @@
+    # The labels do not matter here since they are NOT used during
+    # training.
+    return mixedup_images, labels
+
+
+def preprocess_image(image, label, train=True):
+    image = tf.cast(image, tf.float32) / 255.0
+
+    if train:
+        image = tf.image.resize(image, (BIGGER, BIGGER))
+        image = tf.image.random_crop(image, (RESIZE, RESIZE, 3))
+        image = tf.image.random_flip_left_right(image)
+    else:
+        # Central fraction amount is from here:
+        # https://git.io/J8Kda.
+        image = tf.image.central_crop(image, central_fraction=0.875)
+        image = tf.image.resize(image, (RESIZE, RESIZE))
+
+    return image, label
+
+
+def prepare_dataset(dataset, train=True, batch_size=BATCH_SIZE):
+    if train:
+        dataset = dataset.map(preprocess_image, num_parallel_calls=AUTO)
+        dataset = dataset.shuffle(BATCH_SIZE * 10)
+    else:
+        dataset = dataset.map(
+            lambda x, y: (preprocess_image(x, y, train)), num_parallel_calls=AUTO
+        )
+    dataset = dataset.batch(batch_size)
+
+    if train:
+        dataset = dataset.map(mixup, num_parallel_calls=AUTO)
+
+    dataset = dataset.prefetch(AUTO)
+    return dataset
+
+
+"""
+Note that for brevity, we used mild crops for the training set but in practice
+"Inception-style" preprocessing should be applied. You can refer to
+[this script](https://github.com/sayakpaul/FunMatch-Distillation/blob/main/crop_resize.py)
+for a closer implementation. Also, _**the ground-truth labels are not used during
+training the student.**_
+"""
+
+train_ds = prepare_dataset(train_ds, True)
+validation_ds = prepare_dataset(validation_ds, False)
+test_ds = prepare_dataset(test_ds, False)
+
+"""
+## Visualization
+"""
+
+sample_images, _ = next(iter(train_ds))
+plt.figure(figsize=(10, 10))
+for n in range(25):
+    ax = plt.subplot(5, 5, n + 1)
+    plt.imshow(sample_images[n].numpy())
+    plt.axis("off")
+plt.show()
+
+"""
+## Student model
+
+For the purpose of this example, we will use the standard ResNet50V2 
+([He et al.](https://arxiv.org/abs/1603.05027)).
+"""
+
+
+def get_resnetv2():
+    resnet_v2 = keras.applications.ResNet50V2(
+        weights=None,
+        input_shape=(RESIZE, RESIZE, 3),
+        classes=102,
+        classifier_activation="linear",
+    )
+    return resnet_v2
+
+
+get_resnetv2().count_params()
+
+"""
+Compared to the teacher model, this model has 358 Million less parameters. 
+"""
+
+"""
+## Distillation utility
+
+We will reuse some code from
+[this example](https://keras.io/examples/vision/knowledge_distillation/) on knowledge
+distillation.
+"""
+
+
+class Distiller(tf.keras.Model):
+    def __init__(self, student, teacher):
+        super(Distiller, self).__init__()
+        self.student = student
+        self.teacher = teacher
+
+    def compile(
+        self, optimizer, metrics, distillation_loss_fn, temperature=TEMPERATURE,
+    ):
+        super(Distiller, self).compile(optimizer=optimizer, metrics=metrics)
```

Comment:
You allow arbitrary metrics but don't take them into account in the training step (though you do in the test step). You should add a `self.compiled_metrics.update_state(y, student_predictions)` in `train_step`.

Reply:
Since the labels are not used for calculating any training metrics like accuracy, I didn't do it.

Reply:
Then the code is inconsistent. In your train_step you are returning results for all metrics, not just for the loss. Yet, you are not updating these metrics.

Either you just care about the loss, in which case there should not be other metrics, or you actually support arbitrary metrics, which requires updating them in `train_step` and listing them in `self.metrics`